devdoc/channel.md

```diff
@@ -0,0 +1,267 @@
+
+- **SRS_CHANNEL_43_063: [** signal waiting threads by calling `wake_by_address_all`. **]**
+
+- **SRS_CHANNEL_43_022: [** return `CHANNEL_RESULT_ASYNC`. **]**
+
+**SRS_CHANNEL_43_023: [** If there are any failures, `channel_pull` shall fail and return `CHANNEL_RESULT_ERROR`. **]**
+
+### cancel_pull
+```
+static void cancel_pull(void* context)
+```
+
+`cancel_pull` is the cancel function given to `async_op_create` when `channel_pull` is called.
+
+**SRS_CHANNEL_43_046: [** If the current `EPOCH` and `state` of the `channel` is the same as the initial `EPOCH` and `state` of the `channel` when `channel_pull` was called, `cancel_pull` shall: **]**
+
+- **SRS_CHANNEL_43_047: [** increment the `EPOCH` and set the `state` of the `channel` stored in the given `context` to `CHANNEL_READY` by calling `interlocked_compare_exchange`. **]**
+
+- **SRS_CHANNEL_43_048: [** call the `pull_callback` given to `channel_pull` with `pull_context` as the `pull_context` given to `channel_pull`, `data` as `NULL` and `result` as `CHANNEL_RESULT_CANCELLED`. **]**
+
+
+### channel_push
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_push, THANDLE(CHANNEL), channel, THANDLE(RC_PTR), data, PUSH_CALLBACK, push_callback, void*, push_context, THANDLE(ASYNC_OP)*, out_op_push);
+```
+
+`channel_push` notifies the channel that there is data available and registers the given `push_callback` to be called when the given `data` has been consumed.
+
+**SRS_CHANNEL_43_024: [** If `channel` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_025: [** If `push_callback` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_026: [** If `out_op_push` is `NULL`, `channel_push` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_027: [** `channel_push` shall call `interlocked_add` to obtain the current `state` of the `channel`. **]**
+
+**SRS_CHANNEL_43_028: [** If the current `state` is `CHANNEL_PUSHING` or `CHANNEL_PUSHED`, `channel_push` shall fail and return `CHANNEL_RESULT_REFUSED`. **]**
+
+**SRS_CHANNEL_43_064: [** If the current `state` is `CHANNEL_PULLING`, `channel_push` shall wait for the `state` to change by calling `wait_on_address`. **]**
+
+**SRS_CHANNEL_43_029: [** If the current `state` is `CHANNEL_PULLED`, `channel_push` shall: **]**
+
+- **SRS_WAITER_43_001: [** store `pull_callback` and `pull_context` given to `channel_pull`. **]**
+
+- **SRS_CHANNEL_43_043: [** call `interlocked_compare_exchange` to increment the `EPOCH` and change the state to `CHANNEL_READY`. **]**
+
+- **SRS_CHANNEL_43_030: [** call the stored `pull_callback` with the stored `pull_context`, `data` as the given `data` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+- **SRS_CHANNEL_43_031: [** call the given `push_callback` with `push_context` as the given `push_context` and `result` as `CHANNEL_RESULT_OK`. **]**
```

Comment:
>call the given `push_callback` [](http://example.com/codeflow?start=28&length=30)

I understand the re-entrancy here, and maybe that's a concern for later, but... the document should spell out the mitigations against stack overflows. A producer that pushes to an "always ready" consumer will stack overflow itself.

Reply:
Added note in Reentrancy section